File: articles/0004_std_lib_http_client/src/c/unix_transport.c
#include <httpc/unix_transport.h>
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <sys/un.h>

static Error unix_transport_connect(void* context, const char* host, int port) {
    (void)port;
    UnixClient* self = (UnixClient*)context;
    int sfd;
    struct sockaddr_un addr;

    sfd = self->syscalls->socket(AF_UNIX, SOCK_STREAM, 0);
    if (sfd == -1) {
        return (Error){ErrorType.TRANSPORT, TransportErrorCode.SOCKET_CREATE_FAILURE};
    }

    self->syscalls->memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    self->syscalls->strncpy(addr.sun_path, host, sizeof(addr.sun_path) - 1);

    if (self->syscalls->connect(sfd, (struct sockaddr*)&addr, sizeof(addr)) == -1) {
        self->syscalls->close(sfd);
        return (Error){ErrorType.TRANSPORT, TransportErrorCode.SOCKET_CONNECT_FAILURE};
    }

    self->fd = sfd;
    return (Error){ErrorType.NONE, 0};
}

static Error unix_transport_write(void* context, const void* buffer, size_t len, ssize_t* bytes_written) {
    UnixClient* self = (UnixClient*)context;
    *bytes_written = self->syscalls->write(self->fd, buffer, len);
    if (*bytes_written == -1) {
        return (Error){ErrorType.TRANSPORT, TransportErrorCode.SOCKET_WRITE_FAILURE};
    }
    return (Error){ErrorType.NONE, 0};
}

static Error unix_transport_read(void* context, void* buffer, size_t len, ssize_t* bytes_read) {
    UnixClient* self = (UnixClient*)context;
    *bytes_read = self->syscalls->read(self->fd, buffer, len);
    if (*bytes_read == -1) {
        return (Error){ErrorType.TRANSPORT, TransportErrorCode.SOCKET_READ_FAILURE};
    }
    if (*bytes_read == 0) {
        return (Error){ErrorType.TRANSPORT, TransportErrorCode.CONNECTION_CLOSED};
    }
    return (Error){ErrorType.NONE, 0};
}

static Error unix_transport_close(void* context) {
    UnixClient* self = (UnixClient*)context;
    if (self->fd > 0) {
        int result = self->syscalls->close(self->fd);
        self->fd = 0;
        if (result == -1) {
            return (Error){ErrorType.TRANSPORT, TransportErrorCode.SOCKET_CLOSE_FAILURE};
        }
    }
    return (Error){ErrorType.NONE, 0};
}

TransportInterface* unix_transport_new(const HttpcSyscalls* syscalls_override) {
    if (!default_syscalls_initialized) {
        httpc_syscalls_init_default(&DEFAULT_SYSCALLS);
        default_syscalls_initialized = 1;
    }

    if (!syscalls_override)
    {
        syscalls_override = &DEFAULT_SYSCALLS;
    }

    UnixClient* self = syscalls_override->malloc(sizeof(UnixClient));
    if (!self) {
        return nullptr;
    }
    syscalls_override->memset(self, 0, sizeof(UnixClient));
    self->syscalls = syscalls_override;

    self->interface.context = self;
    self->interface.connect = unix_transport_connect;
    self->interface.write = unix_transport_write;
    self->interface.read = unix_transport_read;
    self->interface.close = unix_transport_close;

    return &self->interface;
}

void unix_transport_destroy(TransportInterface* transport) {
    if (!transport) {
        return;
    }
    UnixClient* self = (UnixClient*)transport;
    self->syscalls->free(self);
}

File: articles/0004_std_lib_http_client/src/rust/src/unix_transport.rs
use crate::error::{Error, Result, TransportError};
use crate::transport::Transport;
use std::io::{Read, Write};
use std::os::unix::net::UnixStream;
use std::net::Shutdown;

#[derive(Default)]
pub struct UnixTransport {
    stream: Option<UnixStream>,
}

impl UnixTransport {
    pub fn new() -> Self {
        Self { stream: None }
    }
}

impl Transport for UnixTransport {
    fn connect(&mut self, path: &str, _port: u16) -> Result<()> {
        match UnixStream::connect(path) {
            Ok(stream) => {
                self.stream = Some(stream);
                Ok(())
            }
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                Err(Error::Transport(TransportError::SocketConnectFailure))
            }
            Err(e) => Err(e.into()),
        }
    }

    fn write(&mut self, buf: &[u8]) -> Result<usize> {
        if let Some(stream) = &mut self.stream {
            let bytes_written = stream.write(buf)?;
            Ok(bytes_written)
        } else {
            Err(Error::Transport(TransportError::SocketWriteFailure))
        }
    }

    fn read(&mut self, buf: &mut [u8]) -> Result<usize> {
        if let Some(stream) = &mut self.stream {
            let bytes_read = stream.read(buf)?;
            if bytes_read == 0 && !buf.is_empty() {
                return Err(Error::Transport(TransportError::ConnectionClosed));
            }
            Ok(bytes_read)
        } else {
            Err(Error::Transport(TransportError::SocketReadFailure))
        }
    }

    fn close(&mut self) -> Result<()> {
        if let Some(stream) = self.stream.take() {
            stream.shutdown(Shutdown::Both)?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use std::sync::atomic::{AtomicUsize, Ordering};
    use super::*;
    use crate::error::{Error, TransportError};
    use std::io::{Read, Write};
    use std::os::unix::net::{UnixListener, UnixStream};
    use std::sync::mpsc;
    use std::thread;
    use std::time::Duration;
    use std::os::unix::io::AsRawFd;

    static TEST_COUNTER: AtomicUsize = AtomicUsize::new(0);

    fn setup_unix_test_server<F>(server_logic: F) -> (String, thread::JoinHandle<()>)
    where
        F: FnOnce(UnixStream) + Send + 'static,
    {
        let count = TEST_COUNTER.fetch_add(1, Ordering::SeqCst);
        let socket_path = format!("/tmp/httpc_rust_test_{}_{}", std::process::id(), count);
        let _ = std::fs::remove_file(&socket_path);

        let listener = UnixListener::bind(&socket_path).unwrap();
        let path_clone = socket_path.clone();

        let handle = thread::spawn(move || {
            if let Ok((stream, _)) = listener.accept() {
                server_logic(stream);
            }
            let _ = std::fs::remove_file(&path_clone);
        });

        (socket_path, handle)
    }

    #[test]
    fn construction_succeeds() {
        let transport = UnixTransport::new();
        assert!(transport.stream.is_none());
    }

    #[test]
    fn connect_succeeds() {
        let (path, server_handle) = setup_unix_test_server(|_stream| {});
        let mut transport = UnixTransport::new();
        assert!(transport.connect(&path, 0).is_ok());
        server_handle.join().unwrap();
    }

    #[test]
    fn write_succeeds() {
        let (tx, rx) = mpsc::channel();
        let msg = "hello unix";
        let (path, server_handle) = setup_unix_test_server(move |mut stream| {
            let mut buffer = String::new();
            stream.read_to_string(&mut buffer).unwrap();
            tx.send(buffer).unwrap();
        });

        let mut transport = UnixTransport::new();
        transport.connect(&path, 0).unwrap();
        transport.write(msg.as_bytes()).unwrap();
        transport.close().unwrap();

        let captured = rx.recv().unwrap();
        assert_eq!(captured, msg);
        server_handle.join().unwrap();
    }

    #[test]
    fn read_succeeds() {
        let msg = "hello from server";
        let (path, server_handle) = setup_unix_test_server(move |mut stream| {
            stream.write_all(msg.as_bytes()).unwrap();
        });

        let mut transport = UnixTransport::new();
        transport.connect(&path, 0).unwrap();

        let mut buf = [0u8; 1024];
        let bytes_read = transport.read(&mut buf).unwrap();

        assert_eq!(bytes_read, msg.len());
        assert_eq!(&buf[..bytes_read], msg.as_bytes());
        server_handle.join().unwrap();
    }

    #[test]
    fn close_succeeds() {
        let (path, handle) = setup_unix_test_server(|_| {});
        let mut transport = UnixTransport::new();
        transport.connect(&path, 0).unwrap();
        assert!(transport.close().is_ok());
        handle.join().unwrap();
    }

    #[test]
    fn close_is_idempotent() {
        let (path, handle) = setup_unix_test_server(|_| {});
        let mut transport = UnixTransport::new();
        transport.connect(&path, 0).unwrap();
        assert!(transport.close().is_ok());
        assert!(transport.close().is_ok());
        handle.join().unwrap();
    }

    #[test]
    fn connect_fails_on_unresponsive_path() {
        let mut transport = UnixTransport::new();
        let path = "/tmp/this-socket-does-not-exist.sock";
        let result = transport.connect(path, 0);
        assert!(result.is_err());
    }

    #[test]
    fn write_fails_on_closed_connection() {
        let (path, handle) = setup_unix_test_server(|stream| {
            let linger = libc::linger { l_onoff: 1, l_linger: 0 };
            unsafe {
                libc::setsockopt(
                    stream.as_raw_fd(),
                    libc::SOL_SOCKET,
                    libc::SO_LINGER,
                    &linger as *const _ as *const libc::c_void,
                    std::mem::size_of::<libc::linger>() as u32,
                );
            }
        });

        let mut transport = UnixTransport::new();
        transport.connect(&path, 0).unwrap();
        handle.join().unwrap();
        thread::sleep(Duration::from_millis(50));

        let result = transport.write(b"fail");
        assert!(result.is_err());
        assert_eq!(result.unwrap_err(), Error::Transport(TransportError::SocketWriteFailure));
    }

    #[test]
    fn read_fails_on_peer_shutdown() {
        let (path, handle) = setup_unix_test_server(|_| {});
        let mut transport = UnixTransport::new();
        transport.connect(&path, 0).unwrap();
        handle.join().unwrap();

        let mut buf = [0u8; 32];
        let result = transport.read(&mut buf);
        assert!(result.is_err());
        assert_eq!(result.unwrap_err(), Error::Transport(TransportError::ConnectionClosed));
    }
}